### PR TITLE
[Inductor] Inplacing with Donated Buffer

### PR DESCRIPTION
**Pull request by BoyuanFeng**
Currently, inductor does not inplace update a buffer if it is an input buffer. Because we don't know if an input will be used by other functions.

Donated buffer provides additional information that an input buffer will not be used by other functions. So we can inplace update donated buffer when possible.

[Dashboard](https://hud.pytorch.org/benchmark/torchbench/inductor_dynamic?dashboard=torchinductor&startTime=Mon,%2011%20Nov%202024%2018:14:36%20GMT&stopTime=Mon,%2018%20Nov%202024%2018:14:36%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(a100)&lBranch=bf/donated-buffer-inplace&lCommit=5df0769c00e6f9000caeb10fd5cbf0b165f69c2a&rBranch=main&rCommit=2b39a8db7741b816b03677a9c6fec1af05640dee)

image
image

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @xmfan

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -1869,6 +1903,20 @@ def _init(self, nodes: List[ir.Operation]) -> None:
             }
         )
 
+    def get_donated_buffer(self) -> Dict[str, SchedulerDonatedBuffer]:
+        name_to_donated_buf = {}
+        for node in self.nodes:
```
Comment (eellison) [inline torch/_inductor/scheduler.py:1908]: Do we need to iterate over every node and every read just to get the Donated Buffers ?

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -2145,6 +2193,21 @@ def add_user(
             for buf in node.get_outputs():
                 buf.set_users(name_to_users[buf.get_name()].items)
 
+        # compute and copy user information of donated buffer
```
Comment (eellison) [inline torch/_inductor/scheduler.py:2196]: Nit: is there a reason to do this as a separate loop instead of where we are adding the other users above ?

File: torch/_inductor/graph.py [hunk 3]
```diff
@@ -990,13 +995,30 @@ def placeholder(
...
         else:
             sizes, strides = self.symbolic_sizes_strides(example)  # type: ignore[assignment]
-        # TODO(jansel): handle input aliasing
-        tensor = TensorBox.create(
-            InputBuffer(
-                name=target,
-                layout=FixedLayout(example.device, example.dtype, sizes, strides),
+
+        tracing_context = torch._guards.TracingContext.try_get()
+        if (
```
Comment (eellison) [inline torch/_inductor/graph.py:1000]: nit: maybe factor out `get_donated_idxs` to utility..

File: test/inductor/test_torchinductor.py [hunk 4]
```diff
@@ -5203,6 +5203,31 @@ def test_layer_norm(self):
         if self.device != "cpu":
             assertGeneratedKernelCountEqual(self, 1)
 
+    def test_matmul_layer_norm(self):
```
Comment (eellison) [inline test/inductor/test_torchinductor.py:5206]: It might be nice in the future to have a way of writing tests for this just in the forward. donated buffer for forward inference is something people occasionally do. 

For now if we factored out `get_donated_idxs` we could patch it and that would give a way of testing.

Could we add test of the form :
```
def foo(x: donated_buffer):
     z = x + 1
     return (z @ x)
```

to make sure we don't inplace the `x + 1` kernel  here ?

File: torch/_inductor/scheduler.py [hunk 5]
```diff
@@ -2145,6 +2193,21 @@ def add_user(
...
+        for node in self.nodes:
+            for read in node.read_writes.reads:
+                if (
+                    read.name in V.graph.graph_inputs
+                    and isinstance(
+                        V.graph.graph_inputs[read.name].data.data, ir.DonatedBuffer  # type: ignore[attr-defined]
+                    )
+                    and not isinstance(read, WeakDep)
+                ):
+                    add_user(read.name, node, node.can_inplace(read))
```
Comment (eellison) [inline torch/_inductor/scheduler.py:2206]: Why do we need this loop ? aren't all these reads already being added in https://github.com/pytorch/pytorch/blob/be51941aae41586429eb888debc3a28fc3961dc9/torch/_inductor/scheduler.py#L2141-L2145?

**Review by eellison — COMMENTED** [review]
Looks good, have some comments. 

What commit were you on when you ran the dashboard link ? Can you be sure to run on a base commit that has been previously benchmarked ?

Also - I saw you had the one skipped test. If there is a known accuracy problem we should fix it prior to turning it on.. 

Also, I wonder if we need to re-run the FakeTensorUpdater after joint graph passes.. prior to this analysis

**Comment by shunting314** [comment]
Nice! Can we try how this impact gpt2 (llm.c) peak memory usage? Back of envelop computation indicates that we can save 3GB.

File: test/inductor/test_distributed_patterns.py [hunk 6]
```diff
@@ -395,6 +395,11 @@ def fn(x):
         self._assert_same_grad(r1, r2)
         self._assert_same_grad(p1, p2)
 
+    # TODO: fix fw_metadata.num_forward in dynamo.
+    # While fn returns two user outputs, dynamo only marks one user outputs
+    # p.sin(). This leads to p being wrongly marked as donated buffer
+    # and inplace updated.
+    @torch._functorch.config.patch(donated_buffer=False)
     def test_nn_param_return3(self):
```
[the remark this answers is not in the document]
Comment (eellison) [reply test/inductor/test_distributed_patterns.py:436]: if it doesn't support it, does it graph break ?  this is valid user code do we give an incorrect result ?

File: torch/_inductor/cudagraph_trees.py [hunk 7]
```diff
@@ -833,6 +833,16 @@ def __init__(
...
 
+        self.live_cudagraph_managed_path_refs: List[Optional[PathOutputIndex]] = [
+            (
+                self._is_alias_of_live_recorded_tensor(t)
+                if isinstance(t, torch.Tensor)
+                else None
+            )
+            for t in inputs
+        ]
+        self.preserved_aliased_inputs: List[bool] = [False] * len(inputs)
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:844]: This should be `InputList`

File: torch/_inductor/cudagraph_trees.py [hunk 8]
```diff
@@ -833,6 +833,16 @@ def __init__(
             if isinstance(t, torch.Tensor) and self._is_cuda_graph_recorded_tensor(t)
         ]
 
+        self.live_cudagraph_managed_path_refs: List[Optional[PathOutputIndex]] = [
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:836]: `InputList`

Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:836]: Can you add a comment why we're doing this?

File: torch/_inductor/ir.py [hunk 9]
```diff
@@ -3692,6 +3692,10 @@ def num_reads(self) -> int:
         return 1
 
 
+class DonatedBuffer(InputBuffer):
```
Comment (eellison) [inline torch/_inductor/ir.py:3695]: docstring ?

**Review by eellison — APPROVED** [review]
🚀 🚀 🚀 

Just one question about the constructing param - if we give incorrect result, would you file an issue or something ? or fix if its easy. 

We can follow up on the llm.c inplacing issue with @shunting314  - not sure what best course of action is

File: test/inductor/test_distributed_patterns.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (anijain2305) [reply test/inductor/test_distributed_patterns.py:436]: I dont fully understand the interaction with this PR. But there is something weird about creating `nn.Parameter` in the compiled frame. There was a concept of sacrificial parameter by @jansel and @yf225.

In this case, we do have a single graph from Dynamo. Just that Dynamo Fx graph does not return the parameter here, it just returns p.sin(). In the modified residual bytecode, Dynamo returns both parameter and p.sin(). So, its all correct. 

https://www.internalfb.com/phabricator/paste/view/P1682708531

[the remark this answers is not in the document]
Comment (eellison) [reply test/inductor/test_distributed_patterns.py:436]: Looks like the donated buffer analysis is not correctly accounting for `torch.ops.aten.set_.source_Tensor(primals_2, add); `. can we fix this before landing @BoyuanFeng

[the remark this answers is not in the document]
Comment (eellison) [reply test/inductor/test_distributed_patterns.py:436]: This would also be fixed if we made aot_autograd return the result of `set_` and drop it, which is something I think we should do anyway. we had another issue in cudagraphs that would have also been fixed by this... cc @bdhirsh

[the remark this answers is not in the document]
Comment (bdhirsh) [reply test/inductor/test_distributed_patterns.py:436]: My recollection is that when we first added the sacrificial parameter logic, we **needed** the set_ to be in the graph in order to let inductor free memory at runtime (and not keep it live til the end of the region).

Maybe a question for @yf225: can/should we just kill this sacrificial parameter code entirely? Or do you think we still need it (for FSDP2 tracing).

I guess one stop-gap alternative if we don't think we can kill it is just to shut off this config whenever dynamo detects a sacrificial parameter in the graph, although that doesn't feel like a great long-term state

[the remark this answers is not in the document]
Comment (yf225) [reply test/inductor/test_distributed_patterns.py]: I think we won't use the sacrificial parameter code for FSDP2 tracing (instead we run 1st iteration with eager to init all parameters in eager, and then in 2nd iteration we start doing compile).

If it's tricky to maintain sacrificial parameter code, I'd vote for graph-breaking on nn.Parameter creation instead.
